### DISCUSSION TITLE
System.Management: unable to get instance of Msvm_VirtualSystemManagementService

### DISCUSSION THREAD

**Discussion by antonalabin** [dotnet/runtime#87095] [Q&A] [answered]
.NET 7.0 project created through [Template Studio for WinUI](https://marketplace.visualstudio.com/items?itemName=TemplateStudio.TemplateStudioForWinUICs)
Package: System.Management

Unable to get instance of Msvm_VirtualSystemManagementService: managementObjectCollection contains no objects.

```c#
ManagementScope scope = new ManagementScope(@"root\virtualization\v2");
using (ManagementClass managementServiceClass = new ManagementClass("Msvm_VirtualSystemManagementService"))
{
    managementServiceClass.Scope = scope;
    ManagementObjectCollection managementObjectCollection = managementServiceClass.GetInstances();
    if (managementObjectCollection.Count == 0)
    {
        throw new ArgumentException("The collection contains no objects", "collection");
    }
}
```

**Comment by danmoseley** [comment]
Can you reduce to simplest standalone repro code? Does it work on .NET Framework?

**Reply by antonalabin** [reply]
Yes. it works with .NET Framework.

**Reply by danmoseley** [reply]
That's odd - the code is essentially unchanged. Any interest in debugging a bit to see where they diverge? Visual Studio should find sources and symbols.

**Reply by danmoseley** [reply]
I pasted the above into a new console app and ran it on my machine for both .NET 7 and .NET Framework 4.8 and both have a count of zero. But, I don't know much about System.Management and whether "Msvm_VirtualSystemManagementService" should be zero for me.

**Reply by antonalabin** [reply]
Do you have Hyper-V enabled on your Windows?

I found that the problem occurs when using a .NET 7.0 project created through [Template Studio for WinUI](https://marketplace.visualstudio.com/items?itemName=TemplateStudio.TemplateStudioForWinUICs). 
If I create a simple .NET 7.0 console project, then there is no problem.

**Reply by danmoseley** [reply]
Is there an [MTA] or [STA] attribute on your Main() ? I wonder whether one of them changes the threading model from the default and that's the reason they behave differently. Console apps are by default MTA.

**Reply by antonalabin** [reply]
Yes, I see the [STA] attribute in App.g.i.cs. 
I can't delete it, because it is restored after a rebuild.

**Reply by danmoseley** [reply]
OK, WPF runs its UI on an STA thread, by design. The fix should be fairly simple: make these calls off the main thread thread. Use `Task.Run(..)` for example.

**Reply by stephentoub** [reply]
> > OK, WPF runs its UI on an STA thread, by design. The fix should be fairly simple: make these calls off the main thread thread. Use `Task.Run(..)` for example.

What's the explanation for "Yes. it works with .NET Framework"?

**Reply by antonalabin** [reply]
> What's the explanation for "Yes. it works with .NET Framework"?

I used a console application with .NET Framework.
Template from Template Studio for WinUI is created only on .NET 7.0.

**Reply by antonalabin** [reply]
> OK, WPF runs its UI on an STA thread, by design. The fix should be fairly simple: make these calls off the main thread thread. Use Task.Run(..) for example.

This does not work.

![image](https://github.com/dotnet/runtime/assets/48159735/b31e637d-fb17-4982-90f2-6a9b65e005a8)

**Reply by danmoseley** [reply]
Can you inspect the thread's GetApartmentState()? And compare the working case

**Comment by antonalabin** [comment, marked as answer]
Problem solved. Required to run as administrator. In the case of WinUI, it is not enough to run Visual Studio with administrator rights and you need to change the project settings.
